File: src/ios/LocalLLMPlugin.swift
import Foundation
import FoundationModels

@objc(LocalLLMPlugin)
class LocalLLMPlugin: CDVPlugin {
    private let implementation = LocalLLM()
    private var availabilityPollingTask: Task<Void, Never>?
    private var forceAvailabilityListenerUpdate = false
    private var listenerCallbackIds: [String: String] = [:]

    // MARK: - Availability

    @objc(systemAvailability:)
    func systemAvailability(command: CDVInvokedUrlCommand) {
        sendOk(
            command,
            payload: ["status": LocalLLM.availability().rawValue]
        )
    }

    @objc(download:)
    func download(command: CDVInvokedUrlCommand) {
        sendError(
            command,
            error: LocalLLMError.featureNotSupported("model download")
        )
    }

    // MARK: - Listeners

    @objc(addAvailabilityListener:)
    func addAvailabilityListener(command: CDVInvokedUrlCommand) {
        guard
            let listenerId = command.argument(at: 0) as? String,
            !listenerId.isEmpty
        else {
            sendError(command, error: LocalLLMError.missingParameter("listenerId"))
            return
        }

        listenerCallbackIds[listenerId] = command.callbackId

        if availabilityPollingTask == nil {
            startAvailabilityPolling()
        } else {
            forceAvailabilityListenerUpdate = true
        }

        pushAvailabilityToListeners(status: LocalLLM.availability().rawValue)
    }

    @objc(removeAvailabilityListener:)
    func removeAvailabilityListener(command: CDVInvokedUrlCommand) {
        guard let listenerId = command.argument(at: 0) as? String else {
            sendOk(command, payload: [:] as [String: Any])
            return
        }
        listenerCallbackIds.removeValue(forKey: listenerId)
        if listenerCallbackIds.isEmpty {
            stopAvailabilityPolling()
        }
        sendOk(command, payload: [:] as [String: Any])
    }

    @objc(removeAllListeners:)
    func removeAllListeners(command: CDVInvokedUrlCommand) {
        listenerCallbackIds.removeAll()
        stopAvailabilityPolling()
        sendOk(command, payload: [:] as [String: Any])
    }

    // MARK: - Sessions & prompts

    @objc(warmup:)
    func warmup(command: CDVInvokedUrlCommand) {
        do {
            let args = command.arguments?.first as? [String: Any] ?? [:]
            guard let sessionId = args["sessionId"] as? String, !sessionId.isEmpty else {
                throw LocalLLMError.missingParameter("sessionId")
            }
            let promptPrefix = args["promptPrefix"] as? String

            try implementation.warmup(sessionId: sessionId, promptPrefix: promptPrefix)
            sendOk(command, payload: [:] as [String: Any])
        } catch {
            sendError(command, error: error)
        }
    }

    @objc(prompt:)
    func prompt(command: CDVInvokedUrlCommand) {
        let options = promptOptions(from: command)

        guard !options.prompt.isEmpty else {
            sendError(command, error: LocalLLMError.missingParameter("prompt"))
            return
        }

        Task {
            do {
                let text = try await implementation.prompt(options: options)
                sendOk(command, payload: ["text": text])
            } catch {
                sendError(command, error: error)
            }
        }
    }

    @objc(endSession:)
    func endSession(command: CDVInvokedUrlCommand) {
        let args = command.arguments?.first as? [String: Any] ?? [:]
        guard let sessionId = args["sessionId"] as? String, !sessionId.isEmpty else {
            sendError(command, error: LocalLLMError.missingParameter("sessionId"))
            return
        }
        implementation.endSession(sessionId)
        sendOk(command, payload: [:] as [String: Any])
    }

    @objc(generateImage:)
    func generateImage(command: CDVInvokedUrlCommand) {
        guard
            let args = command.arguments?.first as? [String: Any],
            let prompt = args["prompt"] as? String,
            !prompt.isEmpty
        else {
            sendError(command, error: LocalLLMError.missingParameter("prompt"))
            return
        }

        let count = max(1, args["count"] as? Int ?? 1)
        let promptImages = args["promptImages"] as? [String] ?? []

        Task {
            do {
                let images = try await implementation.generateImage(
                    prompt: prompt,
                    promptImages: promptImages,
                    variations: count
                )
                sendOk(command, payload: ["pngBase64Images": images])
            } catch {
                sendError(command, error: error)
            }
        }
    }

    // MARK: - Polling

    private func startAvailabilityPolling() {
        guard availabilityPollingTask == nil else { return }

        availabilityPollingTask = Task { [weak self] in
            var lastAvailability: LLMAvailability?
            while !Task.isCancelled {
                guard let self else { return }
                let current = LocalLLM.availability()
                if current != lastAvailability || self.forceAvailabilityListenerUpdate {
                    lastAvailability = current
                    self.forceAvailabilityListenerUpdate = false
                    self.pushAvailabilityToListeners(status: current.rawValue)
                }
                try? await Task.sleep(nanoseconds: 2_000_000_000)
            }
        }
    }

    private func stopAvailabilityPolling() {
        availabilityPollingTask?.cancel()
        availabilityPollingTask = nil
    }

    private func pushAvailabilityToListeners(status: String) {
        let payload: [String: Any] = ["status": status]
        for callbackId in listenerCallbackIds.values {
            guard let result = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: payload) else {
                continue
            }
            result.setKeepCallbackAs(true)
            commandDelegate.send(result, callbackId: callbackId)
        }
    }

    // MARK: - Argument parsing

    private func promptOptions(from command: CDVInvokedUrlCommand) -> LLMPromptOptions {
        let args = command.arguments?.first as? [String: Any] ?? [:]
        return LLMPromptOptions(
            sessionId: args["sessionId"] as? String,
            instructions: args["instructions"] as? String,
            options: llmOptions(from: args["options"] as? [String: Any]),
            prompt: args["prompt"] as? String ?? ""
        )
    }

    private func llmOptions(from dict: [String: Any]?) -> LLMOptions? {
        guard let dict else { return nil }
        return LLMOptions(
            temperature: dict["temperature"] as? Double,
            maximumOutputTokens: dict["maximumOutputTokens"] as? Int
        )
    }

    // MARK: - CDV results

    private func sendOk(_ command: CDVInvokedUrlCommand, payload: [String: Any]) {
        let result = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: payload)
        commandDelegate.send(result, callbackId: command.callbackId)
    }

    private func sendError(_ command: CDVInvokedUrlCommand, error: Error) {
        let nsError = error as NSError

        if #available(iOS 26.0, *), let genError = error as? LanguageModelSession.GenerationError {
            let payload: [String: Any] = [
                "code": "LOCAL_LLM_GENERATION_ERROR",
                "message": genError.localizedDescription,
                "details": String(reflecting: genError),
                "nsErrorDomain": nsError.domain,
                "nsErrorCode": nsError.code,
            ]
            let result = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: payload)
            commandDelegate.send(result, callbackId: command.callbackId)
            return
        }

        // Some FoundationModels errors bridge through as NSError without a typed enum instance.
        if #available(iOS 26.0, *), nsError.domain == "FoundationModels.LanguageModelSession.GenerationError" {
            var payload: [String: Any] = [
                "code": "LOCAL_LLM_GENERATION_ERROR",
                "message": error.localizedDescription,
                "details": String(reflecting: error),
                "nsErrorDomain": nsError.domain,
                "nsErrorCode": nsError.code,
            ]

            if let underlying = nsError.userInfo[NSMultipleUnderlyingErrorsKey] as? [NSError] {
                payload["underlyingErrors"] = underlying.map { u in
                    [
                        "domain": u.domain,
                        "code": u.code,
                        "description": u.localizedDescription,
                        "details": String(reflecting: u),
                    ]
                }
            }

            let result = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: payload)
            commandDelegate.send(result, callbackId: command.callbackId)
            return
        }

        let llmError = error as? LocalLLMError
        let code = llmError?.errorCode ?? "LOCAL_LLM_UNKNOWN_ERROR"
        let message = llmError?.errorDescription ?? error.localizedDescription
        let payload: [String: Any] = [
            "code": code,
            "message": message,
            "details": String(reflecting: error),
            "nsErrorDomain": nsError.domain,
            "nsErrorCode": nsError.code,
        ]
        let result = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: payload)
        commandDelegate.send(result, callbackId: command.callbackId)
    }
}

File: src/android/LocalLLMPlugin.kt
package io.ionic.localllm

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import org.apache.cordova.CallbackContext
import org.apache.cordova.CordovaPlugin
import org.apache.cordova.PluginResult
import org.json.JSONArray
import org.json.JSONObject

class LocalLLMPlugin : CordovaPlugin() {
    private val implementation = LocalLLM()
    private val mainScope = CoroutineScope(Dispatchers.Main + SupervisorJob())
    private val pollingScope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    private var availabilityPollingJob: Job? = null
    private var forceAvailabilityListenerUpdate = false
    private val listenerCallbacks = mutableMapOf<String, CallbackContext>()

    override fun onDestroy() {
        availabilityPollingJob?.cancel()
        pollingScope.cancel()
        mainScope.cancel()
        super.onDestroy()
    }

    override fun execute(
        action: String,
        args: JSONArray,
        callbackContext: CallbackContext,
    ): Boolean {
        when (action) {
            "systemAvailability" -> {
                mainScope.launch {
                    runCatching { implementation.availability() }
                        .onSuccess { status ->
                            callbackContext.success(JSONObject().put("status", status.value))
                        }
                        .onFailure { callbackContext.sendError(it) }
                }
                return true
            }

            "download" -> {
                mainScope.launch {
                    runCatching { implementation.download() }
                        .onSuccess { callbackContext.success() }
                        .onFailure { callbackContext.sendError(it) }
                }
                return true
            }

            "warmup" -> {
                mainScope.launch {
                    runCatching { implementation.warmup() }
                        .onSuccess { callbackContext.success() }
                        .onFailure { callbackContext.sendError(it) }
                }
                return true
            }

            "prompt" -> {
                mainScope.launch {
                    runCatching {
                        val options = parsePromptOptions(args)
                        implementation.prompt(options)
                    }
                        .onSuccess { text ->
                            callbackContext.success(JSONObject().put("text", text))
                        }
                        .onFailure { callbackContext.sendError(it) }
                }
                return true
            }

            "endSession" -> {
                runCatching {
                    val sessionId =
                        args.optJSONObject(0)?.optString("sessionId")
                            ?: throw LocalLLMError.MissingParameter("sessionId")
                    if (sessionId.isBlank()) {
                        throw LocalLLMError.MissingParameter("sessionId")
                    }
                    implementation.endSession(sessionId)
                }
                    .onSuccess { callbackContext.success() }
                    .onFailure { callbackContext.sendError(it) }
                return true
            }

            "generateImage" -> {
                mainScope.launch {
                    runCatching {
                        val payload = args.optJSONObject(0) ?: JSONObject()
                        val prompt =
                            payload.optString("prompt").takeIf { it.isNotBlank() }
                                ?: throw LocalLLMError.MissingParameter("prompt")
                        val count = payload.optInt("count", 1).coerceAtLeast(1)
                        implementation.generateImage(prompt, count)
                    }
                        .onSuccess { images ->
                            callbackContext.success(
                                JSONObject().put("pngBase64Images", JSONArray(images)),
                            )
                        }
                        .onFailure { callbackContext.sendError(it) }
                }
                return true
            }

            "addAvailabilityListener" -> {
                val listenerId =
                    args.optString(0).takeIf { it.isNotBlank() }
                        ?: run {
                            callbackContext.sendError(LocalLLMError.MissingParameter("listenerId"))
                            return true
                        }
                listenerCallbacks[listenerId] = callbackContext
                if (availabilityPollingJob?.isActive != true) {
                    startAvailabilityPolling()
                } else {
                    forceAvailabilityListenerUpdate = true
                }
                pushAvailabilityToListeners()
                return true
            }

            "removeAvailabilityListener" -> {
                val listenerId = args.optString(0)
                listenerCallbacks.remove(listenerId)
                if (listenerCallbacks.isEmpty()) {
                    stopAvailabilityPolling()
                }
                callbackContext.success()
                return true
            }

            "removeAllListeners" -> {
                listenerCallbacks.clear()
                stopAvailabilityPolling()
                callbackContext.success()
                return true
            }
        }

        return false
    }

    private fun startAvailabilityPolling() {
        if (availabilityPollingJob?.isActive == true) return

        availabilityPollingJob =
            pollingScope.launch {
                var lastAvailability: LLMAvailability? = null
                while (isActive) {
                    val current =
                        runCatching { implementation.availability() }
                            .getOrElse { LLMAvailability.Unavailable }

                    if (current != lastAvailability || forceAvailabilityListenerUpdate) {
                        lastAvailability = current
                        forceAvailabilityListenerUpdate = false
                        cordova.activity?.runOnUiThread { pushAvailabilityToListeners(current) }
                    }
                    delay(2000)
                }
            }
    }

    private fun stopAvailabilityPolling() {
        availabilityPollingJob?.cancel()
        availabilityPollingJob = null
    }

    private fun pushAvailabilityToListeners(status: LLMAvailability? = null) {
        val resolved =
            status
                ?: runCatching { implementation.availability() }
                    .getOrElse { LLMAvailability.Unavailable }

        val payload = JSONObject().put("status", resolved.value)
        for ((_, callback) in listenerCallbacks.toList()) {
            val result = PluginResult(PluginResult.Status.OK, payload)
            result.keepCallback = true
            callback.sendPluginResult(result)
        }
    }

    private fun parsePromptOptions(args: JSONArray): LLMPromptOptions {
        val payload = args.optJSONObject(0) ?: JSONObject()
        val prompt =
            payload.optString("prompt").takeIf { it.isNotBlank() }
                ?: throw LocalLLMError.MissingParameter("prompt")

        return LLMPromptOptions(
            sessionId = payload.optString("sessionId").takeIf { it.isNotBlank() },
            instructions = payload.optString("instructions").takeIf { it.isNotBlank() },
            options = parseLLMOptions(payload.optJSONObject("options")),
            prompt = prompt,
        )
    }

    private fun parseLLMOptions(optionsObject: JSONObject?): LLMOptions? {
        if (optionsObject == null) return null

        val temperature =
            optionsObject.optDouble("temperature").takeIf { !it.isNaN() }?.toFloat()?.coerceIn(0f, 1f)
        val maxOutputTokens =
            optionsObject.optInt("maximumOutputTokens").takeIf { it > 0 }?.coerceIn(1, 256)

        if (temperature == null && maxOutputTokens == null) return null
        return LLMOptions(temperature = temperature, maxOutputTokens = maxOutputTokens)
    }

    private fun CallbackContext.sendError(error: Throwable) {
        sendPluginResult(error.toPluginResult())
    }

    private fun Throwable.toPluginResult(): PluginResult {
        val llmError = this as? LocalLLMError
        val payload =
            JSONObject().apply {
                put("code", llmError?.code ?: "LOCAL_LLM_UNKNOWN_ERROR")
                put("message", llmError?.message ?: localizedMessage ?: toString())
                put("details", toString())
            }
        return PluginResult(PluginResult.Status.ERROR, payload)
    }
}
